pennylane/_qubit_device.py

```diff
@@ -873,20 +884,24 @@ def adjoint_jacobian(self, tape):
                 d_op_matrix = operation_derivative(op)
 
             op.inv()
-            phi = self._apply_operation(phi, op)
+            # I'd like to  use op.adjoint() here
+            # then we don't have to re-invert the operation at the end
+            ket = self._apply_operation(ket, op)
 
             if op.grad_method is not None:
                 if param_number in tape.trainable_params:
-                    mu = self._apply_unitary(phi, d_op_matrix, op.wires)
 
-                    jac_column = np.array(
-                        [2 * dot_product_real(lambda_, mu) for lambda_ in lambdas]
-                    )
-                    jac[:, trainable_param_number] = jac_column
+                    ket_temp = self._apply_unitary(ket, d_op_matrix, op.wires)
+
+                    for kk, bra_ in enumerate(bras):
+                        jac[kk, trainable_param_number] = 2 * dot_product_real(bra_, ket_temp)
+
                     trainable_param_number -= 1
                 param_number -= 1
 
-            lambdas = [self._apply_operation(lambda_, op) for lambda_ in lambdas]
+            bras = [self._apply_operation(bra_, op) for bra_ in bras]
             op.inv()
 
+        if return_obs:
+            return expectation, jac
```

Comment:
What is the usecase for `return_obs`? Maybe I missed it, but it doesn't seem to be being used by any of the interfaces.

Reply:
It's not currently used by the interfaces. I wanted to make this function more versatile.  Even though it's not currently used elsewhere in the code base, I think it makes the method itself stronger.